File: rf_ros/hardware_drivers/uhd_usrp/scripts/usrp_driver_node.py
#!/usr/bin/env python3

from __future__ import annotations

import threading

import numpy as np
import rclpy
from rclpy.node import Node

import uhd

from rf_msgs.msg import IqFrame

class UsrpDriverNode(Node):
    def __init__(self) -> None:
        super().__init__("usrp_driver_node")

        self._device_args = str(self.declare_parameter("device_args", "").value)
        self._freq_hz = float(self.declare_parameter("freq_hz", 2.4e9).value)
        self._rate_hz = float(self.declare_parameter("rate_hz", 1e6).value)
        self._gain_db = float(self.declare_parameter("gain_db", 10.0).value)
        self._channel = int(self.declare_parameter("channel", 0).value)
        self._chunk = int(self.declare_parameter("samples_per_chunk", 1024).value)

        self._pub = self.create_publisher(IqFrame, "iq", 10)

        self._thread = threading.Thread(target=self._stream_worker, daemon=True)
        self._thread.start()

    def _stream_worker(self) -> None:
        usrp = uhd.usrp.MultiUSRP(self._device_args)
        usrp.set_rx_rate(self._rate_hz, self._channel)
        usrp.set_rx_freq(uhd.types.TuneRequest(self._freq_hz), self._channel)
        usrp.set_rx_gain(self._gain_db, self._channel)

        fs_hz = float(usrp.get_rx_rate(self._channel))
        fc_hz = float(usrp.get_rx_freq(self._channel))

        st_args = uhd.usrp.StreamArgs("fc32", "sc16")
        st_args.channels = [self._channel]
        streamer = usrp.get_rx_stream(st_args)

        metadata = uhd.types.RXMetadata()
        buffer_samps = streamer.get_max_num_samps()
        samples = np.empty((1, self._chunk), dtype=np.complex64)
        recv_buffer = np.zeros((1, buffer_samps), dtype=np.complex64)

        self.get_logger().info(
            f"USRP driver started: fs={fs_hz / 1e6:.2f} MHz  "
            f"fc={fc_hz / 1e9:.4f} GHz  gain={self._gain_db:.1f} dB  "
            f"chunk={self._chunk} samp"
        )

        stream_cmd = uhd.types.StreamCMD(uhd.types.StreamMode.start_cont)
        stream_cmd.stream_now = True
        streamer.issue_stream_cmd(stream_cmd)

        while True:
            recv_samps = 0
            while recv_samps < self._chunk:
                samps = streamer.recv(recv_buffer, metadata)
                if metadata.error_code != uhd.types.RXMetadataErrorCode.none:
                    print(metadata.strerror())
                if samps:
                    real_samps = min(self._chunk - recv_samps, samps)
                    samples[:, recv_samps : recv_samps + real_samps] = recv_buffer[:, 0:real_samps]
                    recv_samps += real_samps

            try:
                self._publish(samples[self._channel], fs_hz, fc_hz)
            except Exception:
                break

        stop_cmd = uhd.types.StreamCMD(uhd.types.StreamMode.stop_cont)
        streamer.issue_stream_cmd(stop_cmd)
        

    def _publish(self, samples: np.ndarray, fs_hz: float, fc_hz: float) -> bool:

        iq = np.empty(2 * len(samples), dtype=np.float32)
        iq[0::2] = samples.real
        iq[1::2] = samples.imag

        msg = IqFrame()
        msg.stamp = self.get_clock().now().to_msg()
        msg.data = iq.tolist()
        msg.fs_hz = fs_hz
        msg.fc_hz = fc_hz
        self._pub.publish(msg)

    def destroy_node(self) -> None:
        self._thread.join(timeout=3.0)
        super().destroy_node()


def main(args: list[str] | None = None) -> None:
    rclpy.init(args=args)
    node = UsrpDriverNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()


if __name__ == "__main__":
    main()

File: rf_ros/hardware_drivers/uhd_usrp/src/usrp_driver_node.cpp
#include <complex>
#include <string>
#include <vector>

#include <rclcpp/rclcpp.hpp>
#include <rf_msgs/msg/iq_frame.hpp>
#include <uhd/usrp/multi_usrp.hpp>
#include <uhd/stream.hpp>
#include <uhd/types/tune_request.hpp>

#include <boost/thread.hpp>
#include <boost/thread/interruption.hpp>

class UsrpDriverNode : public rclcpp::Node
{
public:
    UsrpDriverNode() : Node("usrp_driver_node")
    {
        device_args_ = declare_parameter<std::string>("device_args", "");
        freq_hz_ = declare_parameter<double>("freq_hz", 2.4e9);
        rate_hz_ = declare_parameter<double>("rate_hz", 1.0e6);
        gain_db_ = declare_parameter<double>("gain_db", 10.0);
        channel_ = static_cast<size_t>(declare_parameter<int>("channel", 0));

        pub_ = create_publisher<rf_msgs::msg::IqFrame>("iq", 10);

        worker_ = boost::thread(&UsrpDriverNode::stream_worker, this);
    }

    ~UsrpDriverNode()
    {
        worker_.interrupt();
        if (worker_.joinable())
            worker_.join();
    }

private:
    void stream_worker()
    {
        auto usrp = uhd::usrp::multi_usrp::make(device_args_);
        usrp->set_rx_rate(rate_hz_, channel_);
        usrp->set_rx_freq(uhd::tune_request_t(freq_hz_), channel_);
        usrp->set_rx_gain(gain_db_, channel_);

        const double fs_hz = usrp->get_rx_rate(channel_);
        const double fc_hz = usrp->get_rx_freq(channel_);

        uhd::stream_args_t st_args("fc32", "sc16");
        st_args.channels = {channel_};
        auto streamer = usrp->get_rx_stream(st_args);

        const size_t buffer_size = streamer->get_max_num_samps();
        std::vector<std::complex<float>> recv_buf(buffer_size);

        RCLCPP_INFO(get_logger(),
                    "USRP driver started: fs=%.2f MHz  fc=%.4f GHz  gain=%.1f dB  buffer=%zu samp",
                    fs_hz / 1e6, fc_hz / 1e9, gain_db_, buffer_size);

        uhd::stream_cmd_t start_cmd(uhd::stream_cmd_t::STREAM_MODE_START_CONTINUOUS);
        start_cmd.stream_now = true;
        streamer->issue_stream_cmd(start_cmd);

        uhd::rx_metadata_t metadata;

        try
        {
            while (true)
            {
                boost::this_thread::interruption_point();
                size_t nsamples = streamer->recv(recv_buf.data(), buffer_size, metadata);
                if (metadata.error_code != uhd::rx_metadata_t::ERROR_CODE_NONE)
                    RCLCPP_WARN(get_logger(), "RX error: %s", metadata.strerror().c_str());
                if (nsamples)
                {
                    // reinterpret complex<float> buffer as interleaved float32 [I0,Q0,I1,Q1,…]
                    const float *samples = reinterpret_cast<const float *>(recv_buf.data());
                    publish(samples, nsamples, fs_hz, fc_hz);
                }
            }
        }
        catch (const boost::thread_interrupted&)
        {
            RCLCPP_INFO(this->get_logger(), "streaming interrupted, exiting");
        }

        uhd::stream_cmd_t stop_cmd(uhd::stream_cmd_t::STREAM_MODE_STOP_CONTINUOUS);
        streamer->issue_stream_cmd(stop_cmd);
    }

    void publish(const float *samples, size_t nsamples, double fs_hz, double fc_hz)
    {
        auto msg = rf_msgs::msg::IqFrame();
        msg.stamp = now();
        msg.fs_hz = fs_hz;
        msg.fc_hz = fc_hz;
        msg.data.assign(samples, samples + nsamples * 2);

        pub_->publish(msg);
    }

    std::string device_args_;
    double freq_hz_;
    double rate_hz_;
    double gain_db_;
    size_t channel_;

    rclcpp::Publisher<rf_msgs::msg::IqFrame>::SharedPtr pub_;
    boost::thread worker_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<UsrpDriverNode>());
    rclcpp::shutdown();
    return 0;
}
